src/main/java/com/snowplowanalytics/redash/RedashClient.java

```diff
@@ -0,0 +1,429 @@
+     */
+    public User getUserById(int userId) throws IOException {
+        String url = baseUrl + USERS_URL_PREFIX + "/" + userId + API_KEY_URL_PARAM + apiKey;
+        String returnValue = get(url, false);
+        return resultResolver(User.class, returnValue);
+    }
+
+    /**
+     * Provides opportunity to retrieve the user-group from redash server by it's id. Implemented as util method.
+     *
+     * @return DataSource instance
+     * @throws IllegalArgumentException if there is no user-group with such id
+     * @throws IOException              if server is not available due to connection error or API key is invalid.
+     */
+    public DataSource getDataSourceById(int id) throws IOException {
+        String url = baseUrl + DATA_SOURCES_URL_PREFIX + "/" + id + API_KEY_URL_PARAM + apiKey;
+        String returnValue = get(url, false);
+        JSONObject jsonObject = new JSONObject(returnValue);
+        if (jsonObject.has(MESSAGE) && MESSAGE_INTERNAL_SERVER_ERROR.equals(jsonObject.getString(MESSAGE))) {
+            throw new IllegalArgumentException(MESSAGE_INTERNAL_SERVER_ERROR);
+        }
+        return new Gson().fromJson(returnValue, DataSource.class);
+    }
+
+    private boolean isDataSourceAlreadyUpToDate(DataSource fromDataBase, DataSource whichToUpdate) {
+        return fromDataBase.getHost().equals(whichToUpdate.getHost())
+                && fromDataBase.getPort() == whichToUpdate.getPort()
+                && fromDataBase.getUser().equals(whichToUpdate.getUser())
+                && fromDataBase.getDbName().equals(whichToUpdate.getDbName());
+    }
+
+    private boolean isEntityAlreadyExists(List<? extends BaseEntity> list, String name) throws IOException {
+        return list.stream().anyMatch(e -> name.equals(e.getName()));
+    }
+
+    private int getIdFromJson(String json) throws JSONException {
+        return new JSONObject(json).getInt(ID);
+    }
+
+    private void checkIfUserGroupExists(int groupId) throws IOException {
+        getUserGroupById(groupId);
+    }
+
+    private <T extends BaseEntity> T resultResolver(Class<T> type, String returnValue) throws IOException {
+        JSONObject jsonObject = new JSONObject(returnValue);
+        if (jsonObject.has(AUTH_TYPE) || jsonObject.has(CREATED_AT)) {
+            return new Gson().fromJson(returnValue, type);
+        }
+        if (jsonObject.getString(MESSAGE).startsWith(MESSAGE_URL_NOT_FOUND) ||
+                jsonObject.getString(MESSAGE).equals(MESSAGE_INTERNAL_SERVER_ERROR)) {
+            throw new IllegalArgumentException(returnValue);
+        }
+        throw new IOException(returnValue);
+    }
+
+    private String post(String url, String json) throws IOException {
+        return post(url, json, false);
+    }
+
+    private String get(String url) throws IOException {
+        return get(url, true);
+    }
+
+    private String delete(String url) throws IOException {
+        return delete(url, false);
```

Comment:
These three abstractions are confusing as they are not consistent - post and delete do no check the response status, but get does. Might be better to remove these and be explicit in each call using an enum for `checkResponseStatus` instead of a boolean which does not have much meaning.